### ISSUE TITLE
Revue de code - Caroline Lebrun-Sabalot

### ISSUE THREAD

**Issue by CaroLebrun** [Sanjopp/MEP_2026_ALAMIDRIDEB_BOULIAIRE_KOUKA_SAMAMA#23]
Bonjour à tous, 
Voici ma revue de code concernant votre projet visant à analyser l'utilité de la batch normalisation et à distinguer si son utilité provient plus du lissage de la fonction de perte que de la stabilisation des activations.

**Organisation globale**
Le travail a bien été réalisé sous Git, avec contribution des 4 membres du projet via un travail sur différentes branches. 

**Présence des fichiers de base d'un repo Git** 
- le README : très complet et clair, il permet de comprendre rapidement les enjeux du projet ainsi que les éléments techniques nécessaires pour se l'approprier et pouvoir le déployer en local ; la table des matières permet de saisir très rapidement les différentes étapes du projet. 
- le .gitignore est complet et conforme aux standards Python : il permet d'exclure les fichiers inutiles, comme les données locales, les caches et les environnements, assurant un dépôt propre adapté au travail collaboratif sous Git. 
- le fichier License : il existe et indique bien que le code est open source. 
- Le projet n'utilise pas un fichier requirements.txt, mais une solution plus moderne et efficace basée sur les fichiers pyproject.toml et uv.lock, qui permettent de définir un environnement reproductible et d’assurer la portabilité du projet.

**Structure du projet**
En remarque préliminaire, on note que la structure du projet (qui a bien été initialisé avec cookiescutter) respecte parfaitement les standards communautaires. 

- Les données
Conformément aux bonnes pratiques, les données ne sont pas stockées dans le dépôt Git mais sur S3. Le dossier data contient donc uniquement des scripts d’import des différents jeux de données, ainsi qu’un README décrivant précisément les sources et les méthodes de chargement, ce qui facilite la gestion de données apparemment volumineuses durant la période développement.

- Les modèles
Les modèles sont rangés dans le dossier models, ce qui permet de centraliser les différentes architectures (MLP, CNN, BatchNorm), et d’améliorer la lisibilité et la maintenabilité du projet.

- Les scripts
Les fonctions principales sont organisées dans le dossier src/training, qui est dédié au pipeline d’entraînement et d’évaluation des modèles.
Chaque script commence par une explication de son rôle et de son utilisation dans le projet, ce qui facilite la compréhension globale. Cette documentation reste cependant parfois un peu succincte pour un lecteur totalement extérieur. La présence de fichiers __init__.py dans src et training permet de structurer correctement le projet en packages Python, et garantit la possibilité d’importer les modules de manière propre et fiable.

- API
Le dossier api contient l’implémentation FastAPI ainsi qu’un module de prédiction chargé d’effectuer l’inférence à partir des modèles entraînés.
Ces fichiers commencent généralement par une brève description de leur rôle, mais la documentation à l'intérieur de chacun de ces fichiers reste limitée, ce qui peut rendre l'appropriation par un lecteur extérieur plus difficile.

- Application Streamlit
Le dossier app contient une interface Streamlit ayant deux fonctionnalités : l’analyse des résultats des expériences, et la mise en œuvre des entraînements via l’API. L’application parait très riche en termes de nombre de fonctions et semble très soignée visuellement, vu la quantité de code dédiée aux paramètres esthétiques. En revanche, elle est uniquement déployée en local, ce qui empêche de la tester directement sans installation. De plus, le code reste peu commenté, ce qui peut rendre sa lecture plus difficile, notamment pour comprendre rapidement la logique des différents blocs et onglets (où les rares commentaires qui servent à marquer la structure n'ont pas la même forme, n'aidant à nouveau pas vraiment à saisir immédiatement l'organisation du code). 

- Déploiement
Le fichier Argo CD permet de relier le dépôt Git à Kubernetes pour automatiser le déploiement de l’application (logique GitOps vue en cours)
L’utilisation de Helm permet de simplifier la gestion des manifests Kubernetes en les rendant paramétrables et réutilisables, notamment pour adapter facilement les configurations du modèle ou de l’environnement.

- CI/CD
Le dossier .github/workflows contient un pipeline CI bien structuré, séparant clairement les étapes de linting (avec Ruff), les tests et la construction de l'image Docker. Il inclut également un template de pull request, ce qui aide à standardiser les contributions et améliore la qualité des reviews.

- Tests
Les tests couvrent à la fois l’API et les modèles, ce qui permet de vérifier leur bon fonctionnement fonctionnel et structurel.
On retrouve notamment des tests sur les formats des entrées/sorties, les gradients, ainsi que la cohérence de la Batch Normalization par rapport à PyTorch, ce qui est particulièrement pertinent dans un contexte de deep learning.

**Conclusion**
Le projet est parfaitement structuré, intéressant, avec une séparation claire des étapes et une chaîne MLOps complète allant de l’entraînement au déploiement. Bravo ! 

Mes pistes d'amélioration sont donc très marginales par rapport à la quantité et la qualité du travail effectué : 
- si la documentation globale est bonne, grâce à un excellent README et des commentaires assez généraux, j'aurais apprécié voir plus de commentaires concernant le fonctionnement interne des fonctions et modules : le code est d'un excellent niveau mais, de ce fait, assez abstrait et difficile à comprendre pour un développeur non senior, surtout en un temps limité. Cela pourrait limiter la compréhension et la maintenabilité à long terme pour un développeur extérieur, ou si le projet est repris longtemps après son développement initial. 
- j'aurais été très curieuse de voir l'interface, mais vu l'ampleur du projet, la revue de code en temps limité ne permet pas de faire soi-même le déploiement en local. Ceci dit, le parcours MLOps ne nécessitait pas de déployer le dashboard, donc c'est plutôt de la curiosité personnelle et du bonus plus qu'une vraie critique du projet, qui à mes yeux respecte intégralement les attentes du cours.
